src/v/config/property_constraints.h

```diff
@@ -0,0 +1,235 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "config/convert.h"
+#include "config/from_string_view.h"
+#include "config/property.h"
+#include "json/_include_first.h"
+#include "json/stringbuffer.h"
+#include "json/writer.h"
+
+#include <seastar/core/sstring.hh>
+#include <seastar/util/noncopyable_function.hh>
+
+#include <fmt/format.h>
+#include <yaml-cpp/node/node.h>
+
+#include <memory>
+#include <string>
+
+namespace cluster {
+struct topic_configuration;
+} // namespace cluster
+
+namespace config {
+
+struct constraint_validator_t {
+    virtual bool validate(const cluster::topic_configuration&) const = 0;
+    virtual void encode_yaml(YAML::Node&) const = 0;
+    virtual void rjson_serialize(json::Writer<json::StringBuffer>&) const = 0;
+    virtual void print(std::ostream&) const = 0;
+    virtual ~constraint_validator_t() = default;
+};
+
+template<typename T>
+struct constraint_validator_range : public constraint_validator_t {
+    using handle_t =
+      typename ss::noncopyable_function<T(const cluster::topic_configuration&)>;
+
+    std::optional<T> min;
+    std::optional<T> max;
+    handle_t handler;
+
+    constraint_validator_range(
+      std::optional<T> min, std::optional<T> max, handle_t handle)
+      : min{std::move(min)}
+      , max{std::move(max)}
+      , handler{std::move(handle)} {}
+
+    virtual bool validate(const cluster::topic_configuration& topic_cfg) const {
+        T topic_val = handler(topic_cfg);
+
+        if (min && *min > topic_val) {
+            return false;
+        }
+
+        if (max && *max < topic_val) {
+            return false;
+        }
+
+        return true;
+    }
+
+    virtual void encode_yaml(YAML::Node& n) const {
+        if (min) {
+            n["min"] = *min;
+        }
+
+        if (max) {
+            n["max"] = *max;
+        }
+    }
+
+    virtual void rjson_serialize(json::Writer<json::StringBuffer>& w) const {
+        w.Key("min");
+        if (min) {
+            w.String(fmt::format("{}", *min));
```

Comment:
note-to-self: writing these at a string is generating weird corner cases in the DT tests. Consider UInt64 for all integral properties.